tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -4340,6 +4340,116 @@ LogicalResult lowerWithScheduleSparseSegmentMeanOpCPU(
+  // memset num_output_elements to 0
+  b.create<memref::StoreOp>(loc,
+                            b.create<arith::ConstantOp>(
+                                loc, b.getIntegerAttr(b.getIntegerType(64), 0)),
+                            num_output_elements, num_output_load_index);
+
+  // for (i = 0; i < num_element; ++i)
+  llvm::SmallVector<scf::ForOp, 4> for_ops;
+  llvm::SmallVector<Value, 4> multidim_index;
+  llvm::SmallVector<Value, 4> multidim_index_casted;
+  for (int i = 0; i < input_rank; ++i) {
+    Value dim = b.create<memref::DimOp>(loc, input, i);
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ dim, /* step */ one);
+    for_ops.push_back(for_op);
+    for_op.getBody()->clear();
+    b.setInsertionPointToStart(for_op.getBody());
+    Value loop_index = for_op.getInductionVar();
+    multidim_index.push_back(loop_index);
+    multidim_index_casted.push_back(
+        b.create<arith::IndexCastOp>(loc, output_elem_type, loop_index));
+  }
+  Value input_elem = b.create<memref::LoadOp>(loc, input, multidim_index);
+  // if (input[i] != 0)
+  Value is_zero;
+  // NOTE: make sure cmp value is the same type as input
+  // otherwise the generated IR will seems ok, however will get segfault
+  // in llvm::CmpInst::CmpInst
+  if (input_elem_type.isa<IntegerType>()) {
+    Value zero_for_cmp =
+        b.create<arith::ConstantOp>(loc, b.getIntegerAttr(input_elem_type, 0));
+    is_zero = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne, input_elem,
+                                      zero_for_cmp);
+  } else if (input_elem_type.isa<FloatType>()) {
+    Value zero_for_cmp =
+        b.create<arith::ConstantOp>(loc, b.getFloatAttr(input_elem_type, 0));
+    is_zero = b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::ONE,
+                                      input_elem, zero_for_cmp);
+  } else {
+    return dominant_op->emitError() << "type other than float or int for "
+                                       "lmhlo_disc::where is not supported yet";
+  }
+  auto if_is_zero = b.create<scf::IfOp>(loc, /*resultTypes*/ llvm::None,
+                                        /* condition */ is_zero,
+                                        /*hasElseRegion*/ false);
+  if_is_zero.getThenRegion().front().clear();
+  b.setInsertionPointToStart(&if_is_zero.getThenRegion().front());
+  Value output_index_count =
+      b.create<memref::LoadOp>(loc, num_output_elements, num_output_load_index);
+  Value output_index_casted =
+      b.create<arith::IndexCastOp>(loc, b.getIndexType(), output_index_count);
+  for (int i = 0; i < input_rank; ++i) {
+    llvm::SmallVector<Value, 2> store_index = {
+        output_index_casted, b.create<arith::ConstantIndexOp>(loc, i)};
+    b.create<memref::StoreOp>(loc, multidim_index_casted[i], index,
+                              store_index);
+  }
+  SmallVector<Value, 1>& num_output_store_index = num_output_load_index;
```

Comment:
Nit: Change `SmallVector<Value, 1>&` to `ArrayRef<Value>`

Reply:
done